File: api/schemas/economics.py
"""Pydantic-контракты REST API экономической модели Cost V2."""
from __future__ import annotations

from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from api.schemas.cost import BlockGeometrySchema
from cost.v2.models import (
    CapacityChoice,
    EconomicScenario,
    MonthlyPlan,
    OperationOverride,
    ReferenceItem,
    ServiceLine,
    SiteConditions,
)


class ReferenceItemSchema(BaseModel):
    code: str = Field(..., min_length=1, max_length=80)
    name: str = Field(..., min_length=1, max_length=300)
    payload: dict[str, Any] = Field(default_factory=dict)
    is_active: bool = True
    valid_from: str | None = None
    valid_to: str | None = None
    source: str = ""
    comment: str = ""
    revision: int = Field(1, ge=1)

    def to_domain(self) -> ReferenceItem:
        return ReferenceItem.from_dict(self.model_dump())


class ReferenceSnapshotSchema(BaseModel):
    revision_id: str
    published_at: str | None = None
    published_by: str
    sections: dict[str, list[ReferenceItemSchema]]
    section_catalog: list[dict[str, str]] = Field(default_factory=list)
    group_catalog: list[dict[str, str]] = Field(default_factory=list)


class ReferenceValidateRequest(BaseModel):
    sections: dict[str, list[ReferenceItemSchema]]


class ReferencePublishRequest(ReferenceValidateRequest):
    base_revision: str
    comment: str = ""


class ReferenceValidationIssueSchema(BaseModel):
    level: Literal["error", "warning"]
    section: str
    code: str
    message: str
    # Пусто, если ошибка относится к записи целиком, а не к отдельному полю.
    field: str = ""


class ReferenceValidationResponse(BaseModel):
    valid: bool
    issues: list[ReferenceValidationIssueSchema]


class TechnicalDriverRequest(BaseModel):
    block: BlockGeometrySchema
    existing_physical: dict[str, Decimal] = Field(default_factory=dict)
    source_id: str | None = None


class TechnicalDriverResponse(BaseModel):
    source_type: Literal["BLAST_GEOMETRY"]
    source_id: str | None = None
    physical: dict[str, Decimal]
    lineage: dict[str, str]


class ReferenceRevisionSchema(BaseModel):
    id: str
    organization_id: str
    sequence_no: int
    published_at: str
    published_by: str
    comment: str


class MonthlyPlanSchema(BaseModel):
    month: str = Field(..., pattern=r"^\d{4}-(0[1-9]|1[0-2])$")
    billed_quantity: Decimal = Field(Decimal("0"), ge=0)
    physical: dict[str, Decimal] = Field(default_factory=dict)
    technical_passport_id: str | None = None

    def to_domain(self) -> MonthlyPlan:
        return MonthlyPlan.from_dict(self.model_dump())


class OperationOverrideSchema(BaseModel):
    operation_code: str
    executor: Literal[
        "OWN", "CUSTOMER", "SUBCONTRACTOR", "THIRD_PARTY_SUPPLIER", "OUT_OF_SCOPE"
    ] = "OWN"
    enabled: bool | None = None
    quantity: Decimal | None = Field(None, ge=0)
    subcontract_rate_rub: Decimal | None = Field(None, ge=0)
    supervision_cost_rub: Decimal = Field(Decimal("0"), ge=0)

    def to_domain(self) -> OperationOverride:
        return OperationOverride.from_dict(self.model_dump())


class SiteConditionsSchema(BaseModel):
    bench_surface_condition_code: str = "PREPARED"
    uncleared_rock_share_pct: Decimal = Field(Decimal("0"), ge=0, le=100)
    drilling_productivity_factor: Decimal = Field(Decimal("1"), gt=0)
    stakeout_mode: Literal[
        "CUSTOMER_CONTROL_POINTS", "CUSTOMER_ALL_HOLES", "CONTRACTOR_ALL_HOLES"
    ] = "CUSTOMER_ALL_HOLES"
    refueling_available: bool = True
    customer_provides_fuel: bool = False
    maintenance_box_available: bool = True
    canteen_available: bool = True
    accommodation_available: bool = True
    meal_cost_rub_person_day: Decimal = Field(Decimal("0"), ge=0)
    accommodation_cost_rub_person_night: Decimal = Field(Decimal("0"), ge=0)
    own_fuel_delivery_cost_rub_trip: Decimal = Field(Decimal("0"), ge=0)
    mobile_maintenance_cost_rub_shift: Decimal = Field(Decimal("0"), ge=0)
    infrastructure_comment: str = ""

    def to_domain(self) -> SiteConditions:
        return SiteConditions.from_dict(self.model_dump())


class ServiceLineSchema(BaseModel):
    id: str
    name: str
    package_code: str
    customer_code: str = ""
    site_code: str = ""
    billing_unit: str
    market_price_rub: Decimal = Field(Decimal("0"), ge=0)
    monthly_plans: list[MonthlyPlanSchema] = Field(default_factory=list)
    operation_overrides: list[OperationOverrideSchema] = Field(default_factory=list)
    site_conditions: SiteConditionsSchema = Field(default_factory=SiteConditionsSchema)
    options: dict[str, Any] = Field(default_factory=dict)
    replaces_service_line_id: str | None = None

    def to_domain(self) -> ServiceLine:
        return ServiceLine.from_dict(self.model_dump())


class CapacityChoiceSchema(BaseModel):
    resource_code: str
    mode: Literal["OVERTIME", "RENT", "SUBCONTRACT", "NEW_ASSET"] = "OVERTIME"
    excess_rate_rub: Decimal = Field(Decimal("0"), ge=0)
    step_capacity: Decimal = Field(Decimal("0"), ge=0)
    step_cost_rub: Decimal = Field(Decimal("0"), ge=0)

    def to_domain(self) -> CapacityChoice:
        return CapacityChoice.from_dict(self.model_dump())


class EconomicScenarioSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str = ""
    name: str = Field(..., min_length=1, max_length=300)
    description: str = ""
    production_unit_code: str = Field(..., min_length=1, max_length=80)
    baseline_service_lines: list[ServiceLineSchema] = Field(default_factory=list)
    candidate_service_lines: list[ServiceLineSchema] = Field(default_factory=list)
    capacity_choices: list[CapacityChoiceSchema] = Field(default_factory=list)
    reference_revision_id: str | None = None

    def to_domain(self) -> EconomicScenario:
        return EconomicScenario.from_dict(self.model_dump())


class StoredScenarioSchema(EconomicScenarioSchema):
    organization_id: str
    created_at: str
    created_by: str
    updated_at: str
    updated_by: str


class CalculationRunSchema(BaseModel):
    id: str
    organization_id: str
    scenario_id: str | None
    reference_revision_id: str
    formula_version: str
    input_snapshot: dict[str, Any]
    result: dict[str, Any]
    created_at: str
    created_by: str
    calculation_scope: Literal["EVENT", "SITE", "UNIT"] = "UNIT"
    technical_passport_id: str | None = None
    site_code: str = ""
    period: str = ""
    technical_formula_version: str = ""


class TechnicalPassportCreateSchema(BaseModel):
    site_code: str = Field(..., min_length=1, max_length=80)
    object_name: str = Field(..., min_length=1, max_length=300)
    previous_passport_id: str | None = None
    reference_revision_id: str | None = None
    formula_version: str = Field("blast-geometry-v1", min_length=1, max_length=80)
    input_snapshot: dict[str, Any] = Field(default_factory=dict)
    selected_variant: dict[str, Any] = Field(default_factory=dict)
    block: BlockGeometrySchema
    existing_physical: dict[str, Decimal] = Field(default_factory=dict)


class TechnicalPassportSchema(BaseModel):
    id: str
    organization_id: str
    site_code: str
    object_name: str
    version_no: int
    previous_passport_id: str | None = None
    reference_revision_id: str
    formula_version: str
    input_snapshot: dict[str, Any]
    selected_variant: dict[str, Any]
    block_snapshot: dict[str, Any]
    physical: dict[str, Decimal]
    lineage: dict[str, str]
    created_at: str
    created_by: str


class EventCalculationRequest(BaseModel):
    technical_passport_id: str
    name: str = Field("Один взрыв", min_length=1, max_length=300)
    production_unit_code: str = Field(..., min_length=1, max_length=80)
    package_code: str = Field(..., min_length=1, max_length=80)
    customer_code: str = ""
    billing_unit: str = "M3"
    billed_quantity: Decimal | None = Field(None, ge=0)
    market_price_rub: Decimal = Field(Decimal("0"), ge=0)
    month: str = Field(..., pattern=r"^\d{4}-(0[1-9]|1[0-2])$")
    operation_overrides: list[OperationOverrideSchema] = Field(default_factory=list)
    site_conditions: SiteConditionsSchema = Field(default_factory=SiteConditionsSchema)
    options: dict[str, Any] = Field(default_factory=dict)
    reference_revision_id: str | None = None

File: frontend/src/types/economics.ts
export type Numeric = number | string;

export type EconomicsReferenceItem = {
  code: string;
  name: string;
  payload: Record<string, unknown>;
  is_active: boolean;
  valid_from: string | null;
  valid_to: string | null;
  source: string;
  comment: string;
  revision: number;
};

export type ReferenceSectionMeta = {
  code: string;
  label: string;
  group: string;
};

export type ReferenceGroupMeta = {
  code: string;
  label: string;
};

export type EconomicsReferenceSnapshot = {
  revision_id: string;
  published_at: string | null;
  published_by: string;
  sections: Record<string, EconomicsReferenceItem[]>;
  section_catalog: ReferenceSectionMeta[];
  group_catalog: ReferenceGroupMeta[];
};

export type ReferenceValidationIssue = {
  level: "error" | "warning";
  section: string;
  code: string;
  message: string;
  /** Поле payload, к которому относится ошибка: форма покажет её под полем. */
  field?: string;
};

export type ReferenceValidation = {
  valid: boolean;
  issues: ReferenceValidationIssue[];
};

export type ReferenceImportResult = {
  file_name: string;
  counts: Record<string, number>;
  sections: Record<string, EconomicsReferenceItem[]>;
};

/** Одно расхождение записи журнала с черновиком: ключ поля и два значения. */
export type PublicFieldChange = {
  key: string;
  old: unknown;
  new: unknown;
};

export type PublicDeltaEntry = {
  /** `new` — записи нет в черновике, `changed`/`deactivated` — есть и расходится. */
  kind: "new" | "changed" | "deactivated";
  section: string;
  public_table: string;
  public_id: number;
  code: string;
  name: string;
  /** Готовая запись справочника: страница подставляет её в черновик как есть. */
  item: EconomicsReferenceItem;
  changes: PublicFieldChange[];
};

export type PublicDelta = {
  /** `false` — журнал project1 недоступен, текст причины в `error`. */
  available: boolean;
  error: string;
  counts: { new: number; changed: number; deactivated: number };
  entries: PublicDeltaEntry[];
};

/** Связь записи справочника со строкой журнала project1. */
export type PublicLink = {
  section: string;
  code: string;
  public_table: string;
  public_id: number;
  synced_at: string | null;
};

export type ReferenceRevision = {
  id: string;
  organization_id: string;
  sequence_no: number;
  published_at: string;
  published_by: string;
  comment: string;
};

export type TechnicalDriverSnapshot = {
  source_type: "BLAST_GEOMETRY";
  source_id: string | null;
  physical: Record<string, Numeric>;
  lineage: Record<string, string>;
};

export type MonthlyEconomicPlan = {
  row_id?: string;
  month: string;
  billed_quantity: Numeric;
  physical: Record<string, Numeric>;
};

export type OperationExecutor =
  | "OWN"
  | "CUSTOMER"
  | "SUBCONTRACTOR"
  | "THIRD_PARTY_SUPPLIER"
  | "OUT_OF_SCOPE";

export type OperationOverride = {
  operation_code: string;
  executor: OperationExecutor;
  enabled: boolean | null;
  quantity: Numeric | null;
  subcontract_rate_rub: Numeric | null;
  supervision_cost_rub: Numeric;
};

export type SiteConditions = {
  bench_surface_condition_code: string;
  uncleared_rock_share_pct: Numeric;
  drilling_productivity_factor: Numeric;
  stakeout_mode:
    | "CUSTOMER_CONTROL_POINTS"
    | "CUSTOMER_ALL_HOLES"
    | "CONTRACTOR_ALL_HOLES";
  refueling_available: boolean;
  customer_provides_fuel: boolean;
  maintenance_box_available: boolean;
  canteen_available: boolean;
  accommodation_available: boolean;
  meal_cost_rub_person_day: Numeric;
  accommodation_cost_rub_person_night: Numeric;
  own_fuel_delivery_cost_rub_trip: Numeric;
  mobile_maintenance_cost_rub_shift: Numeric;
  infrastructure_comment: string;
};

export type EconomicServiceLine = {
  id: string;
  name: string;
  package_code: string;
  customer_code: string;
  site_code: string;
  billing_unit: string;
  market_price_rub: Numeric;
  monthly_plans: MonthlyEconomicPlan[];
  operation_overrides: OperationOverride[];
  site_conditions: SiteConditions;
  options: Record<string, unknown>;
  replaces_service_line_id: string | null;
};

export type CapacityChoice = {
  resource_code: string;
  mode: "OVERTIME" | "RENT" | "SUBCONTRACT" | "NEW_ASSET";
  excess_rate_rub: Numeric;
  step_capacity: Numeric;
  step_cost_rub: Numeric;
};

export type EconomicScenario = {
  id: string;
  name: string;
  description: string;
  production_unit_code: string;
  baseline_service_lines: EconomicServiceLine[];
  candidate_service_lines: EconomicServiceLine[];
  capacity_choices: CapacityChoice[];
  reference_revision_id: string | null;
};

export type StoredEconomicScenario = EconomicScenario & {
  organization_id: string;
  created_at: string;
  created_by: string;
  updated_at: string;
  updated_by: string;
};

export type EconomicMetrics = {
  billed_quantity: number;
  revenue_rub: number;
  variable_cost: number;
  project_direct_cost: number;
  production_cost: number;
  full_internal_cost: number;
  contribution_margin: number;
  project_margin: number;
  production_margin: number;
  full_cost_margin: number;
  cost_market_gap: number;
};

export type ResourceUtilization = {
  month: string;
  resource_code: string;
  resource_name: string;
  demand: number;
  available: number | null;
  utilization_pct: number | null;
  excess: number;
};

export type EconomicPortfolioView = {
  totals: EconomicMetrics;
  periods: Array<EconomicMetrics & { month: string }>;
  service_lines: Array<EconomicMetrics & {
    id: string;
    name: string;
    customer_code: string;
    site_code: string;
    package_code: string;
    billing_unit: string;
    break_even_price_rub: number | null;
  }>;
  resource_utilization: ResourceUtilization[];
  cost_lines: Array<Record<string, string | number>>;
  warnings: string[];
};

export type EconomicCalculationResult = {
  scenario_id: string;
  scenario_name: string;
  production_unit_code: string;
  formula_version: string;
  reference_revision_id: string;
  before: EconomicPortfolioView;
  after: EconomicPortfolioView;
  delta: {
    totals: EconomicMetrics;
    periods: Array<EconomicMetrics & { month: string }>;
    resource_utilization: Array<Record<string, string | number | null>>;
  };
  warnings: string[];
};

export type CalculationRun = {
  id: string;
  organization_id: string;
  scenario_id: string;
  reference_revision_id: string;
  formula_version: string;
  input_snapshot: Record<string, unknown>;
  result: EconomicCalculationResult;
  created_at: string;
  created_by: string;
};
